Add all keyword

fredpi opened the pull request:
Implements #2071

marcelofabri commented on README.md:
```diff
@@ -250,6 +250,22 @@ let noWarning :String = "" // No warning about colons immediately after variable
 
+Including the `all` keyword will disable all rules until the linter sees a matching enable comment:
+
+`// swiftlint:disable all [<rule1> <rule2>...]`
```
should we remove the `[<rule1> <rule2>...]`?

fredpi replied:
I would leave it since it's still possible to exclude single other rules additionally, then disable the all rule and have the single other rules still be in effect.

marcelofabri replied:
Since we already cover disabling individual rules above, I'd just remove them here to make it clearer to understand. I don't expect people to use `all` with other rule identifiers commonly.

fredpi replied:
Makes sense

marcelofabri commented on Source/SwiftLintFramework/Models/Linter.swift:
```diff
@@ -78,11 +78,12 @@ private extension Rule {
         }
 
-        let ruleIDs = Self.description.allIdentifiers +
+        let ruleIds = Self.description.allIdentifiers +
```
is there a reason for renaming this variable?

marcelofabri commented on Source/SwiftLintFramework/Models/Command.swift:
```diff
@@ -84,15 +84,16 @@ public struct Command: Equatable {
+        let ruleTexts = scanner.string.bridge().substring(
+            from: scanner.scanLocation + 1
+        ).components(separatedBy: .whitespaces)
+        ruleIdentifiers = Set(ruleTexts.map { RuleIdentifier($0) })
```
`.map(RuleIdentifier.init)`?

fredpi replied:
This would be ambiguous since there are two inits with string available...

ornithocoder replied:
You can use the qualified symbol name. For instance, `RuleIdentifier.init(_:)` to specify the one without label. This way you can use map - e.g. `.map(RuleIdentifier.init(_:))`.

fredpi replied:
Thanks, implemented that one

marcelofabri commented on Source/SwiftLintFramework/Models/RuleIdentifier.swift:
```diff
@@ -0,0 +1,48 @@
+    private static let allStringRepresentation = "all"
+
+    public var hashValue: Int {
+        switch self {
```
what do you think about using `stringRepresentation.hashValue`?

fredpi replied:
Makes sense

marcelofabri commented on Source/SwiftLintFramework/Models/RuleIdentifier.swift:
```diff
@@ -0,0 +1,48 @@
+    }
+
+    public static func == (lhs: RuleIdentifier, rhs: RuleIdentifier) -> Bool {
+        return lhs.hashValue == rhs.hashValue
```
we shouldn't rely on `hashValue` here. It's better to compare `stringRepresentation`.

fredpi replied:
Makes sense

marcelofabri commented on Tests/SwiftLintFrameworkTests/DisableAllTests.swift:
```diff
@@ -0,0 +1,38 @@
+
+class DisableAllTests: XCTestCase {
+     /// Example violation (identifer_name). Could be replaced with any other single violation.
+    private let violatingPhrase = "let r = 0\n"
```
can you add other violation triggered by other rule in the example? I want to make sure that it disables several ones, just not one.

fredpi replied:
Done

marcelofabri commented on Tests/SwiftLintFrameworkTests/DisableAllTests.swift:
```diff
@@ -0,0 +1,38 @@
+import XCTest
+
+class DisableAllTests: XCTestCase {
+     /// Example violation (identifer_name). Could be replaced with any other single violation.
```
not aligned with the declaration below (one extra whitespace in the beginning)

marcelofabri commented on Source/SwiftLintFramework/Models/Region.swift:
```diff
@@ -12,12 +12,12 @@ import SourceKittenFramework
+    public let disabledRuleIdentifiers: Set<RuleIdentifier>
 
-    public init(start: Location, end: Location, disabledRuleIdentifiers: Set<String>) {
+    public init(start: Location, end: Location, disabledRules: Set<RuleIdentifier>) {
```
I think we should keep the `disabledRuleIdentifiers`

marcelofabri commented on Source/SwiftLintFramework/Models/Region.swift:
```diff
@@ -30,17 +30,17 @@ public struct Region: Equatable {
     public func isRuleDisabled(_ rule: Rule) -> Bool {
         let identifiers = type(of: rule).description.allIdentifiers
-        return !disabledRuleIdentifiers.isDisjoint(with: identifiers)
+        return disabledRuleIdentifiers.contains(.all) ||
```
What do you think about:

```
guard !disabledRuleIdentifiers.contains(.all) else {
   return true
}

return Set(disabledRuleIdentifiers.map { $0.stringRepresentation }).isDisjoint(with: identifiers)
```

fredpi replied:
Rather !isDisjoint... ;)

marcelofabri commented on Tests/SwiftLintFrameworkTests/DisableAllTests.swift:
```diff
@@ -0,0 +1,38 @@
+@testable import SwiftLintFramework
+import XCTest
+
+class DisableAllTests: XCTestCase {
```
I also would like to see tests checking the behavior when `(enable|disable):(next|previous|this)` is used, especially nested with other regions.

marcelofabri commented:
@fredpi Sorry for the delay! Thanks for the contribution, I just made some comments. Let me know what do you think.

fredpi commented:
@marcelofabri Done implementing most of your proposals. For those I didn't implement, I left an explanation why I didn't include it (yet) in the perspective code review section.

ornithocoder commented on CHANGELOG.md:
```diff
@@ -11,6 +11,12 @@
 
+* Add `all` keyword for use in disable / enable statement:
+  `// swiftlint:disable all`.
+  It allows disabling SwiftLint entirely for a particular section.
```
Don't forget to include two trailing spaces at the end of the description (check the other entries on the ChangeLog).

fredpi replied:
Done. I originally did so, but Xcode dropped it automatically ;( Used another editor now

marcelofabri commented on Source/SwiftLintFramework/Models/RuleIdentifier.swift:
```diff
@@ -0,0 +1,42 @@
+
+public enum RuleIdentifier: Hashable, ExpressibleByStringLiteral {
+    case all
+    case some(identifier: String)
```
what do you think about using `ruleID` instead of `some`? 🤔  

Sorry for not catching this earlier.

fredpi replied:
`.some` is actually some legacy naming when the enum was still designed with cases `all` and `.some(let identifiers)` (there were many), so it doesn't fit properly now that it only holds one identifier. Yet I wouldn't go to with `ruleID`which would be confusing to use (`RuleIdentifier.ruleID`). Additionally `ruleID` rather describes the associated value than the case itself.

Since `.some` nonetheless doesn't fit well, I suggest using `case explicit(identifier: String)`

marcelofabri commented:
@fredpi There's only one conflict in the CHANGELOG. Can you please rebase and fix it so we can merge? 💯

fredpi commented:
@marcelofabri I did the rebase and fixed the conflict.

I also changed the RuleIdentifier.explicit case to .single again, which I think, after a bit of consideration, fits better.

marcelofabri commented:
@fredpi Thanks for the amazing contribution! 🎉
